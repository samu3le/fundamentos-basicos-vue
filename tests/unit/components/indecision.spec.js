import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";

describe("Indecision Component", () => {
    let wrapper;
    let clgSpy;
    let getAnswerSpy;

    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () =>
                Promise.resolve({
                    answer: "yes",
                    forced: false,
                    image: "https://yesno.wtf/assets/yes/2.gif",
                }),
        })
    );

    beforeEach(() => {
        wrapper = shallowMount(Indecision);
        clgSpy = jest.spyOn(console, "log");
        getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

        jest.clearAllMocks();
    });

    test("should match with snapshot ", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test("escribir en el input no debe de disparar evento (console.log) ", async() => {
        const input = wrapper.find("input");
        await input.setValue("Hola Mundo");

        expect(clgSpy).toHaveBeenCalledTimes(1);
        // expect(getAnswerSpy).toHaveReturnedTimes(0);
        expect(getAnswerSpy).not.toHaveBeenCalled();
    });

    test('escribir el simbolo "?" debe de disparar el getAnswer', async() => {
        const input = wrapper.find("input");
        await input.setValue("Hola Mundo?");

        expect(getAnswerSpy).toHaveBeenCalled();
    });

    test("pruebas en getAnswer", async() => {
        await wrapper.vm.getAnswer();

        const img = wrapper.find("img");

        expect(img.exists()).toBeTruthy();
        expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
        expect(wrapper.vm.answer).toBe("Si!");
    });

    test("pruebas en el getAnswer - fallo en el API", async() => {
        fetch.mockImplementationOnce(() => Promise.reject("API is down"));
        await wrapper.vm.getAnswer();
        const img = wrapper.find("img");
        expect(img.exists()).toBeFalsy;
        expect(wrapper.vm.answer).toBe("No se pudo cargar el API");
    });
});