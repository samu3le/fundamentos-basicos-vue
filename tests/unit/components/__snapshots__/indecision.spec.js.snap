// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component should match with snapshot  1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
